### PR TITLE
Extend dataloader to also load train data and labels from the database

### PR DESCRIPTION
**Pull request by jo-jstrm**
This does not work yet, because we do not have the option to add IP addresses to the SQLite DB table.

### CODE REVIEW

File: device-identifier/shbdeviceidentifier/commands.py [hunk 1]
```diff
@@ -43,4 +51,38 @@ def read(db: Database, file_path: click.Path, file_type: str):
...
+                    |> range(start: _start, stop: _stop)
+                    |> filter(fn: (r) => r["_measurement"] == "packet")  
+                """
+        train_df, train_labels = DataLoader.from_database(query, params, devices_to_train)
+    else:
+        train_df, train_labels = DataLoader.from_file(
+            training_data_path, training_labels_path, devices_to_train
+        )
+    model = get_model(model_name)
+    model.train(train_df[["data_len", "stream_id"]], train_labels)
```
Comment (TB-DevAcc) [inline device-identifier/shbdeviceidentifier/commands.py:85]: We should probably make sure that there is some kind of callback that shows training progress, otherwise using this from the CLI with a longer training process will be problematic. Should open a new issue for this.

**Review by TB-DevAcc — CHANGES_REQUESTED** [review]
Worked well with in a basic test setup 👍

**Comment by TB-DevAcc** [comment]
We should also merge #89 into this branch before contributing to develop.

File: device-identifier/shbdeviceidentifier/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            "json": DataLoader.labels_from_json,
+        }[get_file_type(training_labels_path)]
+        ip_to_label_map = load_label_lookup(training_labels_path)
+        Y = X["src"].apply(_get_label, args=(ip_to_label_map, devices_to_train)).rename("label")
+        return X, Y
+
+    @staticmethod
+    def from_database(
+        train_data_query: str, params: Dict[str:any], devices_to_train: List[str]
+    ) -> Optional[Tuple[pd.DataFrame, pd.Series]]:
```
[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/dataloader.py:121]: Done

File: device-identifier/shbdeviceidentifier/commands.py [hunk 3]
```diff
@@ -43,4 +51,38 @@ def read(db: Database, file_path: click.Path, file_type: str):
...
+    devices_to_train: List[str] = None,
+):
+    """Train an ML model."""
+    logger.debug(
+        "Will perform training with following parameters:"
+        f"model_name={model_name}, use_database={use_database}, training_data_path={training_data_path}, training_labels_path={training_labels_path}, devices_to_train={devices_to_train}"
+    )
+    if use_database:
+        params = {
+            "_start": datetime.strptime("2022-08-01T11:40:00UTC", "%Y-%m-%dT%H:%M:%S%Z"),
```
[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/commands.py:71]: Forgot about that, sorry.

[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/commands.py:71]: Done. We can set start and stop timestamps for the train data via cli now

File: device-identifier/shbdeviceidentifier/app.py [hunk 4]
```diff
@@ -197,55 +197,17 @@ def query(ctx, data_base, statement_name):
 
 
 @app.command("train")
-@click.argument("model-selector", nargs=1, type=click.STRING)
+@click.argument("model-name", nargs=1, type=click.STRING)
 @click.argument("training-data-path", nargs=1, type=click.STRING)
 @click.argument("training-labels-path", nargs=1, type=click.STRING)
+@click.argument("use-database", nargs=1, type=click.BOOL, default=False)
```
[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/app.py:203]: Would love to have this. I tried implementing `use-database` as an option, but could not get it to work...

File: device-identifier/shbdeviceidentifier/dataloader.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+from .utilities.logging_utilities import spinner
+
+# noinspection PyPep8Naming
+class DataLoader:
+    """
+    Class for loading data from various sources into a pandas DataFrame.
+    """
+
+    @staticmethod
+    def from_influxdb(query: str, params: dict = None) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/dataloader.py:24]: Done

File: device-identifier/shbdeviceidentifier/commands.py (hunk 1 above)
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/commands.py:85]: Sound reasonable. Quick googling indicates that this could become a bit problematic, as training for sklearn differs from training in torch etc. This means that we might have to implement this on the model level.

File: device-identifier/shbdeviceidentifier/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/app.py:204]: Could you try if you can get it to work?

A valid CLI call for testing would be `shbdeviceidentifier --debug train "" "" "" True "2022-08-01 11:38:02" "2022-08-01 11:41:02"`. Make sure to read our pcap into the db first.

File: device-identifier/shbdeviceidentifier/commands.py (hunk 1 above)
Comment (TB-DevAcc) [reply device-identifier/shbdeviceidentifier/commands.py:110]: Yes, MLModel should have some kind of progress method.

Comment (TB-DevAcc) [reply device-identifier/shbdeviceidentifier/commands.py:110]: Is there already an issue for this?

File: device-identifier/shbdeviceidentifier/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TB-DevAcc) [reply device-identifier/shbdeviceidentifier/app.py:204]: I'll take a look today.

File: device-identifier/shbdeviceidentifier/commands.py (hunk 1 above)
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/commands.py:110]: Opened one: #96.

File: device-identifier/shbdeviceidentifier/app.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TB-DevAcc) [reply device-identifier/shbdeviceidentifier/app.py]: I completely changed the implementation to make it possible to use the model as described above. I did not plan on changing everything, that's why I committed to this branch, which looks dumb in hindsight. I'd cherry pick into a new branch if necessary. Anyway, can now be used like this:

```bash
shbdeviceidentifier set-measurement "myData"
shbdeviceidentifier read /path/to/trainingdata.pcap
shbdeviceidentifier read-labels /path/to/traininglabels.json
shbdeviceidentifier train RandomForest
```

Alternatively each command accepts the measurment flag
```bash
shbdeviceidentifier read --measurement "myData" /path/to/trainingdata.pcap
[...]
```

[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/app.py]: The new commands look way better than before. However, I would like to keep ability to set timestamps for splitting the data into train and test sets. With the increasing feature set, also in the GUI, it might might be a good time to discuss the overall workflow of reading data, training a model, and testing/applying the model to unseen data from a user's point of view.

How about we discuss this during the next meeting?

[the remark this answers is not in the document]
Comment (jo-jstrm) [reply device-identifier/shbdeviceidentifier/app.py]: I updated the code a bit. Should work now, together with the other PRs.
